### PR TITLE
Improvements to mindependency dev_requirement conflict resolution

### PR DESCRIPTION
**Pull request by scbedd**
Addresses #37450

The conflict we were running into with `mindependency` is best explained in context of a package.

- `azure-ai-language-conversations` took a relative requirement on `azure-identity` for its tests through the `dev_requirements.txt` file
- `main` azure-identity (eg what is required through dev requirement) required `azure-core>1.31.0`
- `azure-ai-language-conversations` mindependency would install its minimum `azure-core` version of `1.28.0`.
- Attempting to install `azure-core==1.28.0` to meet `mindependency` requirements would CRASH because `azure-identity==1.18.0` being included REQUIRED `azure-core>=1.31.0`.
- Because it's in the dev_requirement, `mindependency` didn't handle it

This PR adds:

- [x] Dependency conflict resolution for the packages defined in dev_requirements
  - If a version is found to be incompatible, it walks backwards through time on pypi until it finds one that IS
- [x] Conflict resolution for a standard version specifier. right now it only handles relative and built requirements. 
  - We now properly leave these alone. Folks, be aware of your dev requirements not breaking mindependency!
- [x] Walk the edge cases a bit further. I'm happy with the fallback of "just leave whatever is there if you don't understand it", that way folks will at least understand the errors if they see them.
- [x] Add tests for `is_package_incompatible` and `resolve_incompatible_package` to cover the edge cases noted in above bullet

### CODE REVIEW

File: tools/azure-sdk-tools/ci_tools/functions.py [hunk 1]
```diff
@@ -625,3 +631,194 @@ def discover_prebuilt_package(dist_directory: str, setup_path: str, package_type
...
+            if package_requirement.key == immutable_requirement.key:
+                # if the dev_req line has a requirement that conflicts with the immutable requirement, we need to resolve it
+                # we KNOW that the immutable requirement will be of form package==version, so we can reliably pull out the version
+                # and check it against the specifier of the dev_req line.
+                immutable_version = next(iter(immutable_requirement.specifier)).version
+
+                if not package_requirement.specifier.contains(immutable_version):
+                    if should_log:
+                        logging.info(
+                            f"Dev req dependency {package_name}'s requirement specifier of {package_requirement} is not compatible with immutable requirement {immutable_requirement}."
```
Comment (mccoyp) [inline tools/azure-sdk-tools/ci_tools/functions.py:665]: Nit: can we run `black` on this file? Some of these lines look long

File: tools/azure-sdk-tools/ci_tools/functions.py [hunk 2]
```diff
@@ -625,3 +631,194 @@ def discover_prebuilt_package(dist_directory: str, setup_path: str, package_type
...
+                        )
+                    return False
+
+    return True
+
+
+def resolve_compatible_package(
+    package_name: str, package_version: Optional[str], immutable_requirements: List[Requirement]
```
Comment (mccoyp) [inline tools/azure-sdk-tools/ci_tools/functions.py:673]: `package_version` doesn't seem to be used here?

File: eng/tox/install_depend_packages.py [hunk 3]
```diff
@@ -101,8 +101,15 @@ def install_dependent_packages(setup_py_file_path, dependency_type, temp_dir):
         override_added_packages.extend(check_pkg_against_overrides(pkg_spec))
 
     logging.info("%s released packages: %s", dependency_type, released_packages)
-    # filter released packages from dev_requirements and create a new file "new_dev_requirements.txt"
-    dev_req_file_path = filter_dev_requirements(setup_py_file_path, released_packages, temp_dir, dependency_type)
+
+    additionalFilterFn = None
```
[the remark this answers is not in the document]
Comment (scbedd) [reply eng/tox/install_depend_packages.py:105]: I slipped into `js` for a second 👍

File: tools/azure-sdk-tools/ci_tools/functions.py (hunk 1 above)
Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/functions.py:665]: I thought I did I'll re-run it 👍

File: tools/azure-sdk-tools/ci_tools/functions.py (hunk 2 above)
Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/functions.py:673]: Agree. Will factor it out before merge.

File: tools/azure-sdk-tools/ci_tools/functions.py (hunk 1 above)
Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/functions.py:665]: image

apparently not? I think I need to add newlines to longer strings myself.

**Review by mccoyp — APPROVED** [review]
I think this looks solid! Just a few more instances of long lines and JS naming conventions, but the logic makes sense to me

**Comment by scbedd** [comment]
> Member

Annoyed I missed the last `additionalFilterFn` when I updated it everywhere else. Thanks for the review @mccoyp I fixed the last one.
